File: app/src/main/java/com/gandhasiri/app/TreeRepository.java
package com.gandhasiri.app;

import android.app.Application;
import androidx.lifecycle.LiveData;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TreeRepository {
    private TreeDao treeDao;
    private LiveData<List<Tree>> allTrees;
    private final ExecutorService executorService = Executors.newFixedThreadPool(4);

    public TreeRepository(Application application) {
        AppDatabase db = AppDatabase.getDatabase(application);
        treeDao = db.treeDao();
        allTrees = treeDao.getAllTrees();
    }

    public LiveData<List<Tree>> getAllTrees() {
        return allTrees;
    }

    public LiveData<Tree> getTreeById(int id) {
        return treeDao.getTreeById(id);
    }

    public void insertWithLog(Tree tree, double initialGirth) {
        executorService.execute(() -> {
            long id = treeDao.insert(tree);
            treeDao.insertGrowthLog(new GrowthLog((int)id, initialGirth, System.currentTimeMillis(), "Healthy"));
        });
    }

    public void updateTreeWithLog(Tree tree, double newGirth) {
        executorService.execute(() -> {
            tree.setGirth(newGirth);
            treeDao.update(tree);
            treeDao.insertGrowthLog(new GrowthLog(tree.getId(), newGirth, System.currentTimeMillis(), "Healthy"));
        });
    }

    public void delete(Tree tree) {
        executorService.execute(() -> treeDao.delete(tree));
    }

    public LiveData<List<GrowthLog>> getGrowthLogs(int treeId) {
        return treeDao.getGrowthLogsForTree(treeId);
    }
}

File: app/src/main/java/com/gandhasiri/app/TreeRepository.kt
package com.gandhasiri.app

import android.app.Application
import androidx.lifecycle.LiveData
import java.util.concurrent.Executors

class TreeRepository(application: Application) {
    private val treeDao: TreeDao
    private val allTrees: LiveData<List<Tree>>
    private val executorService = Executors.newFixedThreadPool(4)

    init {
        val db = AppDatabase.getDatabase(application)
        treeDao = db.treeDao()
        allTrees = treeDao.getAllTrees()
    }

    fun getAllTrees(): LiveData<List<Tree>> = allTrees

    fun getTreeById(id: Int): LiveData<Tree> = treeDao.getTreeById(id)

    fun insertWithLog(tree: Tree, initialGirth: Double) {
        executorService.execute {
            val id = treeDao.insert(tree)
            treeDao.insertGrowthLog(GrowthLog(treeId = id.toInt(), girth = initialGirth, timestamp = System.currentTimeMillis(), healthStatus = "Healthy"))
        }
    }

    fun updateTreeWithLog(tree: Tree, newGirth: Double) {
        executorService.execute {
            tree.girth = newGirth
            treeDao.update(tree)
            treeDao.insertGrowthLog(GrowthLog(treeId = tree.id, girth = newGirth, timestamp = System.currentTimeMillis(), healthStatus = "Healthy"))
        }
    }

    fun delete(tree: Tree) {
        executorService.execute { treeDao.delete(tree) }
    }

    fun getGrowthLogs(treeId: Int): LiveData<List<GrowthLog>> = treeDao.getGrowthLogsForTree(treeId)
}
